les/peer.go

```diff
@@ -585,102 +593,93 @@ func (p *peer) Handshake(td *big.Int, head common.Hash, headNum uint64, genesis
 	send = send.add("headHash", head)
 	send = send.add("headNum", headNum)
 	send = send.add("genesisHash", genesis)
-	if server != nil {
-		// Add some information which services server can offer.
-		if !server.config.UltraLightOnlyAnnounce {
-			send = send.add("serveHeaders", nil)
-			send = send.add("serveChainSince", uint64(0))
-			send = send.add("serveStateSince", uint64(0))
-
-			// If local ethereum node is running in archive mode, advertise ourselves we have
-			// all version state data. Otherwise only recent state is available.
-			stateRecent := uint64(core.TriesInMemory - 4)
-			if server.archiveMode {
-				stateRecent = 0
-			}
-			send = send.add("serveRecentState", stateRecent)
-			send = send.add("txRelay", nil)
-		}
-		send = send.add("flowControl/BL", server.defParams.BufLimit)
-		send = send.add("flowControl/MRR", server.defParams.MinRecharge)
 
-		var costList RequestCostList
-		if server.costTracker.testCostList != nil {
-			costList = server.costTracker.testCostList
-		} else {
-			costList = server.costTracker.makeCostList(server.costTracker.globalFactor())
-		}
-		send = send.add("flowControl/MRC", costList)
-		p.fcCosts = costList.decode(ProtocolLengths[uint(p.version)])
-		p.fcParams = server.defParams
-
-		// Add advertised checkpoint and register block height which
-		// client can verify the checkpoint validity.
-		if server.oracle != nil && server.oracle.isRunning() {
-			cp, height := server.oracle.stableCheckpoint()
-			if cp != nil {
-				send = send.add("checkpoint/value", cp)
-				send = send.add("checkpoint/registerHeight", height)
-			}
-		}
-	} else {
-		// Add some client-specific handshake fields
-		p.announceType = announceTypeSimple
-		if p.trusted {
-			p.announceType = announceTypeSigned
-		}
-		send = send.add("announceType", p.announceType)
+	// If the protocol version is beyond les4, then pass the forkID
+	// as well. Check http://eips.ethereum.org/EIPS/eip-2124 for more
+	// spec detail.
+	if p.version >= lpv4 {
+		send = send.add("forkID", forkID)
 	}
-
-	recvList, err := p.sendReceiveHandshake(send)
+	// Add client-specified or server-specified fields
+	if sendCallback != nil {
+		sendCallback(&send)
+	}
+	// Exchange the handshake packet and resolve the received one.
+	recvList, err := p.exchangeHandshake(send)
 	if err != nil {
 		return err
 	}
 	recv, size := recvList.decode()
-	if p.rejectUpdate(size) {
+	if size > allowedUpdateBytes {
```

Comment:
This works too but why didn't you like the original logic that handles handshake announcements similarly to in-protocol announcements? Both of them use bandwidth similarly so I think the same cumulative limitation can apply to both. Of course if there is a good reason why you think it is better this way then we can change it.

Reply:
updated